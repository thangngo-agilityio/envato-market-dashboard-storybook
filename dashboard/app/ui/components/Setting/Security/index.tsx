'use client';

import { memo, useCallback, useState } from 'react';
import { Controller } from 'react-hook-form';
import { ViewOffIcon, ViewIcon } from '@chakra-ui/icons';

import {
  Box,
  VStack,
  useDisclosure,
  Text,
  Button,
  useToast,
  Heading,
  FormControl,
  FormLabel,
  Flex,
} from '@chakra-ui/react';

// Constants
import {
  AUTH_SCHEMA,
  ERROR_MESSAGES,
  IMAGES,
  STATUS,
  SUCCESS_MESSAGES,
} from '@/lib/constants';

// Hooks
import { useForm, useUpdatePassword } from '@/lib/hooks';

// Components
import { FallbackImage, InputField } from '@/ui/components';

// Stores
import { authStore } from '@/lib/stores';
import { TPassword } from '@/lib/interfaces';

// Utils
import { customToast } from '@/lib/utils';
import { QueryProvider } from '@/ui/providers';

const SecurityPage = () => {
  const { mutate: updatePassword, isPending } = useUpdatePassword();
  const user = authStore((state) => state.user);

  const toast = useToast();
  const {
    control,
    formState: {
      errors: { root },
    },
    watch,
    handleSubmit,
    reset,
    clearErrors,
  } = useForm<TPassword>({
    defaultValues: {
      oldPassword: '',
      newPassword: '',
      memberId: user?.id,
    },
    mode: 'onBlur',
  });

  const handleClearErrorMessage = useCallback(
    (field: keyof TPassword, onChange: (value: string) => void) =>
      (data: string) => {
        clearErrors(field);
        onChange(data);
      },
    [clearErrors],
  );

  const handleUpdatePasswordSuccess = useCallback(() => {
    toast(
      customToast(
        SUCCESS_MESSAGES.UPDATE_SUCCESS.title,
        SUCCESS_MESSAGES.UPDATE_SUCCESS.description,
        STATUS.SUCCESS,
      ),
    );
    reset();
  }, [reset, toast]);

  const handleSubmitForm = useCallback(
    (updatedInfo: TPassword) => {
      updatePassword(updatedInfo, {
        onSuccess: handleUpdatePasswordSuccess,
        onError: () => {
          toast(
            customToast(
              ERROR_MESSAGES.OLD_PASSWORD_INCORRECT,
              ERROR_MESSAGES.UPDATE_FAIL.description,
              STATUS.ERROR,
            ),
          );
        },
      });
    },
    [handleUpdatePasswordSuccess, toast, updatePassword],
  );

  const { isOpen: isShowOldPassword, onToggle: onShowOldPassWord } =
    useDisclosure();
  const { isOpen: isShowNewPassword, onToggle: onShowNewPassWord } =
    useDisclosure();

  const renderPasswordIcon = useCallback(
    (isCorrect: boolean, callback: typeof onShowOldPassWord): JSX.Element => {
      const Icon = isCorrect ? ViewIcon : ViewOffIcon;

      return (
        <Icon
          color="gray.400"
          w="25px"
          h="25px"
          cursor="pointer"
          onClick={callback}
        />
      );
    },
    [],
  );

  const [isSubmit] = useState<boolean>(false);
  const isDisabledSubmitBtn: boolean =
    isSubmit || !Object.values(watch()).every((value) => value) || isPending;

  return (
    <Flex w="full" gap={6} direction="row" alignItems="center">
      <VStack
        mt={6}
        as="form"
        gap={6}
        onSubmit={handleSubmit(handleSubmitForm)}
        id="register-form"
        w="full"
        alignItems="flex-start"
        flex={3}
      >
        <Box alignContent="start">
          <Heading
            as="h3"
            textTransform="capitalize"
            color="text.quinary"
            fontSize="2xl"
            fontWeight="bold"
            mb={3}
          >
            Password
          </Heading>
          <Text fontSize="14px" color="text.ternary">
            Change or view your password
          </Text>
        </Box>
        <Controller
          rules={AUTH_SCHEMA.OLD_PASSWORD}
          control={control}
          name="oldPassword"
          render={({ field: { onChange, ...rest }, fieldState: { error } }) => {
            const { message } = error ?? {};

            return (
              <FormControl>
                <FormLabel
                  color="text.ternary"
                  fontWeight="medium"
                  fontSize="sm"
                  mb={3}
                >
                  Old password
                </FormLabel>
                <InputField
                  type={isShowOldPassword ? 'text' : 'password'}
                  rightIcon={renderPasswordIcon(
                    isShowOldPassword,
                    onShowOldPassWord,
                  )}
                  {...rest}
                  isError={!!message}
                  errorMessages={message}
                  isDisabled={isPending}
                  onChange={handleClearErrorMessage('oldPassword', onChange)}
                  aria-label="oldPassword"
                  role="textbox"
                />
              </FormControl>
            );
          }}
        />

        <Controller
          control={control}
          rules={AUTH_SCHEMA.NEW_PASSWORD}
          name="newPassword"
          render={({ field: { onChange, ...rest }, fieldState: { error } }) => (
            <FormControl>
              <FormLabel
                color="text.ternary"
                fontWeight="medium"
                fontSize="sm"
                mb={3}
              >
                New password
              </FormLabel>
              <InputField
                type={isShowNewPassword ? 'text' : 'password'}
                rightIcon={renderPasswordIcon(
                  isShowNewPassword,
                  onShowNewPassWord,
                )}
                {...rest}
                isDisabled={isPending}
                onChange={handleClearErrorMessage('newPassword', onChange)}
                role="textbox"
                aria-label="newPassword"
              />
              <Text color="text.ternary" fontSize="xs">
                Minimum 8 characters
              </Text>

              {!!error && (
                <Text
                  fontSize="sm"
                  color="danger.300"
                  fontWeight="normal"
                  pos="absolute"
                >
                  {error?.message}
                </Text>
              )}
            </FormControl>
          )}
        />

        <Box mb={7} alignSelf="end">
          <Text color="red" textAlign="center" py={2} h={10}>
            {root?.message}
          </Text>
          <Flex>
            <Button
              type="submit"
              aria-label="btn-save-change"
              px={4}
              textTransform="capitalize"
              form="register-form"
              isDisabled={isDisabledSubmitBtn}
              w="none"
            >
              Save Change
            </Button>
          </Flex>
        </Box>
      </VStack>
      <FallbackImage
        src={IMAGES.PASSWORD.url}
        alt={IMAGES.PASSWORD.alt}
        objectFit="contain"
        w={265}
        h={265}
        display={{ base: 'none', xl: 'block' }}
        flex={4}
      />
    </Flex>
  );
};

const WrappedUserForm = () => (
  <QueryProvider>
    <SecurityPage />
  </QueryProvider>
);

const Security = memo(WrappedUserForm);
export default Security;
