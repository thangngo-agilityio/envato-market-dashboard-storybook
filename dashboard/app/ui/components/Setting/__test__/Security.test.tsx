// Component

import Security from '../Security';

const { render } = testLibReactUtils;
describe('FaqPage', () => {
  it('renders FaqPage component correctly', () => {
    const { asFragment } = render(<Security />);

    expect(asFragment()).toMatchSnapshot();
  });
});
