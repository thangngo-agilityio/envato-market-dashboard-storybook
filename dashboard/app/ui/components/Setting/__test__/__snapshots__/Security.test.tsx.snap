// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FaqPage renders FaqPage component correctly 1`] = `
<DocumentFragment>
  <div
    class="css-2u6dg0"
  >
    <form
      class="chakra-stack css-otx4ua"
      id="register-form"
    >
      <div
        class="css-zf62g1"
      >
        <h3
          class="chakra-heading css-1kyw7bp"
        >
          Password
        </h3>
        <p
          class="chakra-text css-1hfnimn"
        >
          Change or view your password
        </p>
      </div>
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-5rqtjg"
          for="field-:r2:"
          id="field-:r2:-label"
        >
          Old password
        </label>
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <div
            class="chakra-input__group css-1091msq"
            data-group="true"
          >
            <input
              aria-label="oldPassword"
              class="chakra-input css-y1ysar"
              id="field-:r3:"
              name="oldPassword"
              role="textbox"
              type="password"
              value=""
            />
            <button
              aria-label="The eye icon"
              class="chakra-input__right-element css-1stj06r"
              data-testid="right-icon-input"
              type="button"
            >
              <svg
                class="chakra-icon css-xnnqpn"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g
                  fill="currentColor"
                >
                  <path
                    d="M23.2,10.549a20.954,20.954,0,0,0-4.3-3.6l4-3.995a1,1,0,1,0-1.414-1.414l-.018.018a.737.737,0,0,1-.173.291l-19.5,19.5c-.008.007-.018.009-.026.017a1,1,0,0,0,1.631,1.088l4.146-4.146a11.26,11.26,0,0,0,4.31.939h.3c4.256,0,8.489-2.984,11.051-5.8A2.171,2.171,0,0,0,23.2,10.549ZM16.313,13.27a4.581,4.581,0,0,1-3,3.028,4.3,4.3,0,0,1-3.1-.19.253.253,0,0,1-.068-.407l5.56-5.559a.252.252,0,0,1,.407.067A4.3,4.3,0,0,1,16.313,13.27Z"
                  />
                  <path
                    d="M7.615,13.4a.244.244,0,0,0,.061-.24A4.315,4.315,0,0,1,7.5,12,4.5,4.5,0,0,1,12,7.5a4.276,4.276,0,0,1,1.16.173.244.244,0,0,0,.24-.062l1.941-1.942a.254.254,0,0,0-.1-.421A10.413,10.413,0,0,0,12,4.75C7.7,4.692,3.4,7.7.813,10.549a2.15,2.15,0,0,0-.007,2.9,21.209,21.209,0,0,0,3.438,3.03.256.256,0,0,0,.326-.029Z"
                  />
                </g>
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-5rqtjg"
          for="field-:r4:"
          id="field-:r4:-label"
        >
          New password
        </label>
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <div
            class="chakra-input__group css-1091msq"
            data-group="true"
          >
            <input
              aria-label="newPassword"
              class="chakra-input css-y1ysar"
              id="field-:r5:"
              name="newPassword"
              role="textbox"
              type="password"
              value=""
            />
            <button
              aria-label="The eye icon"
              class="chakra-input__right-element css-1stj06r"
              data-testid="right-icon-input"
              type="button"
            >
              <svg
                class="chakra-icon css-xnnqpn"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g
                  fill="currentColor"
                >
                  <path
                    d="M23.2,10.549a20.954,20.954,0,0,0-4.3-3.6l4-3.995a1,1,0,1,0-1.414-1.414l-.018.018a.737.737,0,0,1-.173.291l-19.5,19.5c-.008.007-.018.009-.026.017a1,1,0,0,0,1.631,1.088l4.146-4.146a11.26,11.26,0,0,0,4.31.939h.3c4.256,0,8.489-2.984,11.051-5.8A2.171,2.171,0,0,0,23.2,10.549ZM16.313,13.27a4.581,4.581,0,0,1-3,3.028,4.3,4.3,0,0,1-3.1-.19.253.253,0,0,1-.068-.407l5.56-5.559a.252.252,0,0,1,.407.067A4.3,4.3,0,0,1,16.313,13.27Z"
                  />
                  <path
                    d="M7.615,13.4a.244.244,0,0,0,.061-.24A4.315,4.315,0,0,1,7.5,12,4.5,4.5,0,0,1,12,7.5a4.276,4.276,0,0,1,1.16.173.244.244,0,0,0,.24-.062l1.941-1.942a.254.254,0,0,0-.1-.421A10.413,10.413,0,0,0,12,4.75C7.7,4.692,3.4,7.7.813,10.549a2.15,2.15,0,0,0-.007,2.9,21.209,21.209,0,0,0,3.438,3.03.256.256,0,0,0,.326-.029Z"
                  />
                </g>
              </svg>
            </button>
          </div>
        </div>
        <p
          class="chakra-text css-1iathfz"
        >
          Minimum 8 characters
        </p>
      </div>
      <div
        class="css-szawqj"
      >
        <p
          class="chakra-text css-rj1t1b"
        />
        <div
          class="css-k008qs"
        >
          <button
            aria-label="btn-save-change"
            class="chakra-button css-11exfvr"
            disabled=""
            form="register-form"
            type="submit"
          >
            Save Change
          </button>
        </div>
      </div>
    </form>
    <img
      alt="This is password image"
      class="chakra-image css-1623zui"
      src="images/reset-password.svg"
    />
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
